src/geometry/manifold/manifold.lean

```diff
@@ -0,0 +1,539 @@
+                           class
+`atlas H M`              : when `M` is a manifold modelled on `H`, the atlas of this manifold
+                           structure, i.e., the set of charts
+`diffeomorph G M M'`     : the set of diffeomorphisms between the manifolds M and M' for the
+                           groupoid G
+
+As a basic example, we give the instance
+`instance manifold_model_space (H : Type*) [topological_space H] : manifold H H`
+saying that a topological space is a manifold over itself, with the identity as unique chart. This
+manifold structure is compatible with any groupoid.
+
+## Implementation notes
+
+The atlas in a manifold is *not* a maximal atlas in general: the notion of maximality depends on the
+groupoid one considers, and changing groupoids changes the maximal atlas. With the current
+formalization, it makes sense first to choose the atlas, and then to ask whether this precise atlas
+defines a smooth manifold, an orientable manifold, and so on.
+
+We use the letter `H` for the model space thinking of the case of manifolds with boundary, where the
+model space is a half space.
+-/
+
+noncomputable theory
+local attribute [instance, priority 0] classical.decidable_inhabited classical.prop_decidable
+
+universes u
+
+variables {H : Type u} {M : Type*} {M' : Type*} {M'' : Type*}
+
+local infixr  ` ≫ₕ `:100 := local_homeomorph.trans
+
+open set local_homeomorph
+
+section groupoid
+
+/- One could add to the definition of a structure groupoid the fact that the restriction of an
+element of the groupoid to any open set still belongs to the groupoid.
+(This is in Kobayashi-Nomizu.)
+I am not sure I want this, for instance on H × E where E is a vector space, and the groupoid is made
+of functions respecting the fibers and linear in the fibers (so that a manifold over this groupoid
+is naturally a vector bundle) I prefer that the members of the groupoid are always defined on
+sets of the form s × E
+
+The only nontrivial requirement is locality: if a local homeomorphisms belongs to the groupoid
+around each point in its domain of definition, then it belongs to the groupoid. Without this
+requirement, the composition of diffeomorphisms does not have to be a diffeomorphism.
+
+We also require that being a member of the groupoid only depends on the values on the source, as
+the other values are irrelevant.
+-/
+/-- A structure groupoid is a set of local homeomorphisms of a topological space stable under
+composition and inverse. They appear in the definition of the smoothness class of a manifold. -/
+structure structure_groupoid (H : Type u) [topological_space H] :=
+(members      : set (local_homeomorph H H))
+(comp         : ∀e e' : local_homeomorph H H, e ∈ members → e' ∈ members → e ≫ₕ e' ∈ members)
+(inv          : ∀e : local_homeomorph H H, e ∈ members → e.symm ∈ members)
+(id_mem       : local_homeomorph.refl H ∈ members)
+(locality     : ∀e : local_homeomorph H H, (∀x ∈ e.source, ∃s, is_open s ∧
```

Comment:
Neighborhoods are in general not open, but local homeomorphisms are defined on open sets, so restricting `e` to a non-open neighborhood of `x` would create more problems than it would solve (technically, this would be the restriction of `e` to the interior of the neighborhood...).